### PR TITLE
region_t bindings

### PR DESCRIPTION
**Pull request by jasperrxzhu**
To work around the lack of `char*` support by pybind11, I created a new `PyReg` class that acts as a Python wrapper for `region_t` structures. Previously with the C++ API, users had to allocate their own arrays for timestamps and data, and then invoke `init_region` to initialize a `region_t` structure properly. In this case, I have not created bindings for `ival_t` or `char*` and a Python user is not responsible for allocating the above arrays. Instead, the user specifies the desired size (in terms of number of elements in the stream) and a schema for the data stream, and the `PyReg` object takes care of memory management with respect to timestamps and data. 

To write data to `region_t` structure underlying a `PyReg` object, I've added a `write_data` binding. This is essentially the equivalent of using fetch and then dereferencing to write data to a `region_t` object. I make use of pybind11's Python C++ interface (https://pybind11.readthedocs.io/en/stable/advanced/pycpp/index.html) and the `PyReg` object's schema to make sure that the Python value that a user wants to write to the stream is converted the appropriate underlying C++ data format. 

Bindings for all other `region_t` related functions are created as methods for the `PyReg` class. 

I have a `Cvtr` class because I initially expected to need more utility functions, but these can just be moved into the DataType class if you'd like (i.e. in tilt/base/type.h).

### CODE REVIEW

**Comment by jasperrxzhu** [comment]
PR is still incomplete.

Currently looking for an alternative to the non-const reference issue being raised in `python/region/src/bindings.cpp:19`. There is also a bug that does not allow me to compile/execute multiple TiLT queries sequentially within the same python file.

File: include/tilt/pass/codegen/llvmgen.h [hunk 1]
```diff
@@ -25,13 +25,25 @@ extern const char* vinstr_str;
...
     LLVMGenCtx(const LoopNode* loop, llvm::LLVMContext* llctx) :
         IRGenCtx(nullptr, &loop->syms, new map<Sym, llvm::Value*>()),
         loop(loop), llctx(llctx), map_backup(unique_ptr<map<Sym, llvm::Value*>>(out_sym_tbl))
     {}
 
+    /* ctor should only be used with getStructPadding */
+    LLVMGenCtx(void) :
+        IRGenCtx(nullptr, nullptr, nullptr),
+        loop(nullptr), llctx(nullptr), map_backup(nullptr)
+    {}
```
Comment (anandj91) [inline include/tilt/pass/codegen/llvmgen.h]: Why do we need two ctors? there is another one defined that accepts LLVMContext as argument at line 64.

File: python/src/bindings/ir.cpp [hunk 2]
```diff
@@ -118,9 +108,9 @@ PYBIND11_MODULE(ir, m) {
         .value("gt", MathOp::GT)
         .value("gte", MathOp::GTE)
         .value("eq", MathOp::EQ)
-        .value("not", MathOp::NOT)
-        .value("and", MathOp::AND)
-        .value("or", MathOp::OR);
+        .value("_not", MathOp::NOT)
+        .value("_and", MathOp::AND)
+        .value("_or", MathOp::OR);
```
Comment (anandj91) [inline python/src/bindings/ir.cpp:113]: We need to keep the naming style consistent. Can we make all the function binding to start with "_" ? Or is there a better way to make it consistent. I had similar conflict issue with C++ APIs as well. That is why I started using this convention.

File: src/pass/codegen/llvmgen.cpp [hunk 3]
```diff
@@ -534,6 +535,38 @@ unique_ptr<llvm::Module> LLVMGen::Build(const Loop loop, llvm::LLVMContext& llct
...
+    LLVMGen llgen(&llctx);
+    return llgen.build_struct_padding(dtype);
+}
+
+StructPaddingInfo LLVMGen::build_struct_padding(const DataType& dtype)
+{
+    llvm::Type* dtllvm = lltype(dtype);
+    auto total_bytes = llmod()->getDataLayout().getTypeSizeInBits(dtllvm).getFixedSize();
+    ASSERT(total_bytes % 8 == 0);
+    total_bytes = total_bytes / 8;
```
Comment (anandj91) [inline src/pass/codegen/llvmgen.cpp]: Can you use llsizeof() for this? You can either modify llsizeof to return an integer or convert the ConstantInt llvm value back to an integer.

File: src/pass/codegen/llvmgen.cpp [hunk 4]
```diff
@@ -534,6 +535,38 @@ unique_ptr<llvm::Module> LLVMGen::Build(const Loop loop, llvm::LLVMContext& llct
...
 }
 
+StructPaddingInfo LLVMGen::GetStructPadding(const DataType& dtype)
+{
+    llvm::LLVMContext& llctx = ExecEngine::Get()->GetCtx();
+    LLVMGen llgen(&llctx);
+    return llgen.build_struct_padding(dtype);
+}
+
+StructPaddingInfo LLVMGen::build_struct_padding(const DataType& dtype)
```
Comment (anandj91) [inline src/pass/codegen/llvmgen.cpp]: Can we name StructPaddingInfo to just PaddingInfo? It looks like you are creating this object even if the datatype is not a struct.

File: python/include/pyreg.h [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+                                const tilt::StructPaddingInfo& padding_info,
+                                char* fetch);
+    template<typename T>
+    void append_btype_data_to_sstream(std::ostringstream &os, char *fetch);
+
+    /* PyReg::write_data helpers */
+    void write_data_to_ptr(py::object payload,
+                           const tilt::DataType& dt,
+                           const tilt::StructPaddingInfo& padding_info,
+                           char* raw_data_ptr);
```
Comment (anandj91) [inline python/include/pyreg.h]: do we need these helper functions to be part of the class? You can define them only in the pyreg.cpp file

File: python/include/pyreg.h [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    template<typename T>
+    void write_btype_data_to_ptr(py::object payload, char* raw_data_ptr);
```
Comment (anandj91) [inline python/include/pyreg.h]: same for this.

File: python/include/pyreg.h [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+class PyReg {
+public:
+    PyReg(idx_t size,
+          tilt::DataType schema);
+    ~PyReg();
+    PyReg(const PyReg&) = delete;
+
+    region_t* get_reg(void);
+    std::string str(void);
+    void write_data(py::object payload, ts_t t, idx_t i);
```
Comment (anandj91) [inline python/include/pyreg.h:25]: we also need a function to read data from the region. We can add that later.

**Review by anandj91 — CHANGES_REQUESTED** [review]

**Review by anandj91 — APPROVED** [review]

**Pull request merged**